app/controllers/admin/deploy_groups_controller.rb

```diff
@@ -82,50 +82,87 @@ def create_all_stages_preview
+    preexisting_stages, = stages_for_creation
+    template_stages = deploy_group.environment.template_stages.all
+
+    preexisting_stages.each do |stage|
+      template_stage = template_stages.detect { |ts| ts.project_id == stage.project.id }
+      merge_stage(stage, template_stage)
     end
 
     redirect_to [:admin, deploy_group]
   end
 
+  def self.create_all_stages(deploy_group)
+    _, missing_stages = stages_for_creation(deploy_group)
+    missing_stages.each do |template_stage|
+      create_stage_with_group(template_stage, deploy_group)
+    end
+  end
+
   private
 
-  # returns a list of stages already created and list of stages to create (through their template stages)
+  def merge_stage(stage, template_stage)
+    return unless template_stage
+    return if template_stage.deploy_groups.include?(stage.deploy_groups.first)
+    return if stage.is_template
+    return if stage.deploy_groups.count == 0
+    return if stage.deploy_groups.count > 1
+
+    template_stage.deploy_groups += stage.deploy_groups
+    template_stage.next_stage_ids.delete(stage.id)
+    template_stage.save!
+
+    stage.project.stages.reload # need to reload to make verify_not_part_of_pipeline have current data and not fail
+    stage.soft_delete!
+  end
+
   def stages_for_creation
-    environment = deploy_group.environment
-    template_stages = environment.template_stages.all
-    deploy_group_stages = deploy_group.stages.all
-
-    preexisting_stages = []
-    missing_stages = []
-    Project.where(include_new_deploy_groups: true).each do |project|
-      template_stage = template_stages.detect { |ts| ts.project_id == project.id }
-      deploy_group_stage = deploy_group_stages.detect { |dgs| dgs.project.id == project.id }
-      if deploy_group_stage
-        preexisting_stages << deploy_group_stage
-      elsif template_stage
-        missing_stages << template_stage
+    self.class.stages_for_creation(deploy_group)
+  end
+
+  class << self
+    # returns a list of stages already created and list of stages to create (through their template stages)
+    def stages_for_creation(deploy_group)
+      environment = deploy_group.environment
+      template_stages = environment.template_stages.all
+      deploy_group_stages = deploy_group.stages.all
+
+      preexisting_stages = []
+      missing_stages = []
+      Project.where(include_new_deploy_groups: true).each do |project|
+        template_stage = template_stages.detect { |ts| ts.project_id == project.id }
+        deploy_group_stage = deploy_group_stages.detect { |dgs| dgs.project.id == project.id }
+        if deploy_group_stage
+          preexisting_stages << deploy_group_stage
+        elsif template_stage
+          missing_stages << template_stage
+        end
       end
+
+      [preexisting_stages, missing_stages]
     end
 
-    [preexisting_stages, missing_stages]
-  end
+    def create_stage_with_group(template_stage, deploy_group)
+      stage = Stage.build_clone(template_stage)
```

Comment:
What I'm suggesting is that cloning a stage should not clone the is_template value since this means every regular clone will copy it and that most new stages will have `is_template` which means we have lots of templates by accident